File: skills/model-switch/scripts/switch.py
#!/usr/bin/env python3
# Model Switch - Wechsle zwischen free, premium und auto Modellen
# Ersetzt ALLE openrouter/* Eintraege durch das Zielmodell

import json
import sys
import os
import subprocess
import shutil
from datetime import datetime

CONFIG_PATH = os.path.join(os.path.expanduser("~"), ".openclaw", "openclaw.json")
BACKUP_PATH = os.path.join(os.path.expanduser("~"), ".openclaw", "openclaw.json.backup")

# Modell-Kennungen
FREE_MODEL = "openrouter/openrouter/free"
PREMIUM_MODEL = "openrouter/minimax/minimax-m2.5"
AUTO_MODEL = "openrouter/auto"

MODEL_MAP = {
    "free": FREE_MODEL,
    "premium": PREMIUM_MODEL,
    "auto": AUTO_MODEL,
}

LABEL_MAP = {
    FREE_MODEL: "Free",
    PREMIUM_MODEL: "Premium",
    AUTO_MODEL: "Auto",
}


def replace_all_openrouter_models(obj, target_model):
    """Rekursiv ALLE openrouter/* Werte in model/primary Keys ersetzen."""
    count = 0
    if isinstance(obj, dict):
        for key, value in obj.items():
            if key in ("model", "primary") and isinstance(value, str) and value.startswith("openrouter/") and value != target_model:
                obj[key] = target_model
                count += 1
            elif isinstance(value, (dict, list)):
                count += replace_all_openrouter_models(value, target_model)
    elif isinstance(obj, list):
        for i, item in enumerate(obj):
            if isinstance(item, (dict, list)):
                count += replace_all_openrouter_models(item, target_model)
    return count


def find_current_models(obj):
    """Zaehle alle openrouter/* Modelle in model/primary Keys."""
    found = {}
    def _scan(node):
        if isinstance(node, dict):
            for key, value in node.items():
                if key in ("model", "primary") and isinstance(value, str) and value.startswith("openrouter/"):
                    found[value] = found.get(value, 0) + 1
                elif isinstance(value, (dict, list)):
                    _scan(value)
        elif isinstance(node, list):
            for item in node:
                if isinstance(item, (dict, list)):
                    _scan(item)
    _scan(obj)
    return found


def restart_gateway():
    """Gateway neu starten - Windows-kompatibel."""
    candidates = [
        os.path.join(os.environ.get("ProgramFiles", r"C:\Program Files"), "nodejs", "openclaw.cmd"),
        shutil.which("openclaw") or "",
    ]
    for path in candidates:
        if path and os.path.exists(path):
            try:
                result = subprocess.run(
                    [path, "gateway", "restart"],
                    capture_output=True, text=True, timeout=20
                )
                if result.returncode == 0:
                    return "OK"
                else:
                    return f"FEHLER: {result.stderr.strip()}"
            except Exception as e:
                return f"FEHLER: {e}"
    return "openclaw nicht gefunden - manuell: openclaw gateway restart"


def main():
    if len(sys.argv) < 2:
        print("Usage: switch.py free|premium|auto|status")
        sys.exit(1)

    command = sys.argv[1].lower()

    if not os.path.exists(CONFIG_PATH):
        print(f"FEHLER: Config nicht gefunden: {CONFIG_PATH}")
        sys.exit(1)

    with open(CONFIG_PATH, 'r', encoding='utf-8') as f:
        config = json.load(f)

    # --- STATUS ---
    if command == "status":
        found = find_current_models(config)
        if not found:
            print("=== MODEL STATUS ===")
            print("Keine openrouter/* Modelle gefunden.")
            sys.exit(0)

        most_common = max(found, key=found.get)
        label = LABEL_MAP.get(most_common, most_common)
        print(f"=== MODEL STATUS ===")
        print(f"Aktuelles Modell: {label} ({most_common})")
        for m, c in found.items():
            print(f"  {m}: {c}x")
        sys.exit(0)

    # --- SWITCH ---
    if command not in MODEL_MAP:
        print(f"Unbekannter Befehl: {command}")
        print("Verfuegbar: free | premium | auto | status")
        sys.exit(1)

    target_model = MODEL_MAP[command]
    label = LABEL_MAP[target_model]

    # Bereits auf Zielmodell?
    found = find_current_models(config)
    most_common = max(found, key=found.get) if found else None
    if most_common == target_model and len(found) == 1:
        print(f"Bereits auf {label} ({target_model})")
        sys.exit(0)

    # Backup
    try:
        shutil.copy(CONFIG_PATH, BACKUP_PATH)
    except Exception as e:
        print(f"Backup-Warnung: {e}")

    # ALLE openrouter/* durch Zielmodell ersetzen
    total_changed = replace_all_openrouter_models(config, target_model)

    if total_changed == 0:
        print(f"Keine Eintraege zum Aendern gefunden.")
        sys.exit(0)

    # Config schreiben
    with open(CONFIG_PATH, 'w', encoding='utf-8') as f:
        json.dump(config, f, indent=2, ensure_ascii=False)

    print(f"=== MODEL GEWECHSELT ===")
    print(f"Neues Modell: {label} ({target_model})")
    print(f"Eintraege geaendert: {total_changed}")

    # Gateway restart
    gw_status = restart_gateway()
    print(f"Gateway restart: {gw_status}")
    print(f"Zeit: {datetime.now().strftime('%H:%M:%S')}")


if __name__ == "__main__":
    main()

File: skills/model-switch/scripts/switch.js
#!/usr/bin/env node
/**
 * Model Switch - Wechsle zwischen free, premium und auto Modellen
 * Node.js Version - keine Python Abhaengigkeit
 * 
 * Ersetzt ALLE openrouter/* Eintraege durch das Zielmodell
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const CONFIG_PATH = path.join(process.env.USERPROFILE || process.env.HOME, '.openclaw', 'openclaw.json');
const BACKUP_PATH = path.join(process.env.USERPROFILE || process.env.HOME, '.openclaw', 'openclaw.json.backup');

// Modell-Kennungen
const FREE_MODEL = 'openrouter/openrouter/free';
const PREMIUM_MODEL = 'openrouter/minimax/minimax-m2.5';
const AUTO_MODEL = 'openrouter/auto';

const MODEL_MAP = {
    'free': FREE_MODEL,
    'premium': PREMIUM_MODEL,
    'auto': AUTO_MODEL,
};

const LABEL_MAP = {
    [FREE_MODEL]: 'Free',
    [PREMIUM_MODEL]: 'Premium',
    [AUTO_MODEL]: 'Auto',
};

function replaceAllOpenrouterModels(obj, targetModel) {
    let count = 0;
    if (Array.isArray(obj)) {
        for (let i = 0; i < obj.length; i++) {
            if (typeof obj[i] === 'object' && obj[i] !== null) {
                count += replaceAllOpenrouterModels(obj[i], targetModel);
            }
        }
    } else if (typeof obj === 'object' && obj !== null) {
        for (const key of Object.keys(obj)) {
            if ((key === 'model' || key === 'primary') && 
                typeof obj[key] === 'string' && 
                obj[key].startsWith('openrouter/') && 
                obj[key] !== targetModel) {
                obj[key] = targetModel;
                count++;
            } else if (typeof obj[key] === 'object' && obj[key] !== null) {
                count += replaceAllOpenrouterModels(obj[key], targetModel);
            }
        }
    }
    return count;
}

function findCurrentModels(obj) {
    const found = {};
    
    function scan(node) {
        if (Array.isArray(node)) {
            for (const item of node) {
                if (typeof item === 'object' && item !== null) {
                    scan(item);
                }
            }
        } else if (typeof node === 'object' && node !== null) {
            for (const key of Object.keys(node)) {
                if ((key === 'model' || key === 'primary') && 
                    typeof node[key] === 'string' && 
                    node[key].startsWith('openrouter/')) {
                    const model = node[key];
                    found[model] = (found[model] || 0) + 1;
                } else if (typeof node[key] === 'object' && node[key] !== null) {
                    scan(node[key]);
                }
            }
        }
    }
    
    scan(obj);
    return found;
}

function restartGateway() {
    const openclawCmd = path.join(process.env.ProgramFiles || 'C:\\Program Files', 'nodejs', 'openclaw.cmd');
    
    // Try openclaw.cmd first
    if (fs.existsSync(openclawCmd)) {
        try {
            execSync(`"${openclawCmd}" gateway restart`, { timeout: 20000, stdio: 'pipe' });
            return 'OK';
        } catch (e) {
            return `FEHLER: ${e.message}`;
        }
    }
    
    // Try npx openclaw
    try {
        execSync('npx openclaw gateway restart', { timeout: 20000, stdio: 'pipe' });
        return 'OK';
    } catch (e) {
        return `FEHLER: ${e.message}`;
    }
}

function main() {
    const args = process.argv.slice(2);
    const command = args[0]?.toLowerCase();
    
    if (!command) {
        console.log('Usage: switch.js free|premium|auto|status');
        console.log('');
        console.log('Commands:');
        console.log('  free     - Wechsle zu openrouter/openrouter/free');
        console.log('  premium  - Wechsle zu openrouter/minimax/minimax-m2.5');
        console.log('  auto     - Wechsle zu openrouter/auto');
        console.log('  status   - Zeige aktuelles Modell');
        process.exit(1);
    }
    
    // Check config exists
    if (!fs.existsSync(CONFIG_PATH)) {
        console.log(`FEHLER: Config nicht gefunden: ${CONFIG_PATH}`);
        process.exit(1);
    }
    
    let config;
    try {
        const content = fs.readFileSync(CONFIG_PATH, 'utf8');
        config = JSON.parse(content);
    } catch (e) {
        console.log(`FEHLER: Config konnte nicht gelesen werden: ${e.message}`);
        process.exit(1);
    }
    
    // STATUS
    if (command === 'status') {
        const found = findCurrentModels(config);
        const keys = Object.keys(found);
        
        if (keys.length === 0) {
            console.log('=== MODEL STATUS ===');
            console.log('Keine openrouter/* Modelle gefunden.');
            process.exit(0);
        }
        
        // Find most common
        let mostCommon = keys[0];
        for (const m of keys) {
            if (found[m] > found[mostCommon]) {
                mostCommon = m;
            }
        }
        
        const label = LABEL_MAP[mostCommon] || mostCommon;
        
        console.log('=== MODEL STATUS ===');
        console.log(`Aktuelles Modell: ${label} (${mostCommon})`);
        for (const [model, count] of Object.entries(found)) {
            const l = LABEL_MAP[model] || model;
            console.log(`  ${l}: ${model} (${count}x)`);
        }
        process.exit(0);
    }
    
    // Validate command
    if (!MODEL_MAP[command]) {
        console.log(`Unbekannter Befehl: ${command}`);
        console.log('Verfuegbar: free | premium | auto | status');
        process.exit(1);
    }
    
    const targetModel = MODEL_MAP[command];
    const label = LABEL_MAP[targetModel];
    
    // Already on target model?
    const found = findCurrentModels(config);
    const keys = Object.keys(found);
    let mostCommon = keys.length > 0 ? keys[0] : null;
    
    if (mostCommon === targetModel && keys.length === 1) {
        console.log(`Bereits auf ${label} (${targetModel})`);
        process.exit(0);
    }
    
    // Backup
    try {
        fs.copyFileSync(CONFIG_PATH, BACKUP_PATH);
    } catch (e) {
        console.log(`Backup-Warnung: ${e.message}`);
    }
    
    // Replace all openrouter/* with target model
    const totalChanged = replaceAllOpenrouterModels(config, targetModel);
    
    if (totalChanged === 0) {
        console.log('Keine Eintraege zum Aendern gefunden.');
        process.exit(0);
    }
    
    // Write config
    try {
        fs.writeFileSync(CONFIG_PATH, JSON.stringify(config, null, 2), 'utf8');
    } catch (e) {
        console.log(`FEHLER: Config konnte nicht geschrieben werden: ${e.message}`);
        process.exit(1);
    }
    
    console.log('=== MODEL GEWECHSELT ===');
    console.log(`Neues Modell: ${label} (${targetModel})`);
    console.log(`Eintraege geaendert: ${totalChanged}`);
    
    // Gateway restart
    const gwStatus = restartGateway();
    console.log(`Gateway restart: ${gwStatus}`);
    
    const now = new Date();
    console.log(`Zeit: ${now.getHours().toString().padStart(2,'0')}:${now.getMinutes().toString().padStart(2,'0')}:${now.getSeconds().toString().padStart(2,'0')}`);
}

main();